src/clr-core/internal/base/focus-trap.base.ts

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
+ * This software is released under MIT license.
+ * The full license information can be found in LICENSE in the root directory of this project.
+ */
+
+import { html, LitElement } from 'lit-element';
+import { property } from '../decorators/property';
+import { FocusTrapTracker } from '../services/focus-trap-tracker.service';
+import { addAttributeValue, isHTMLElement, removeAttributeValue } from '../utils/dom';
+import { isBrowser } from '../utils/exists';
+
+export class CdsBaseFocusTrap extends LitElement {
+  private previousFocus: Element;
+  private topReboundEl: HTMLElement;
+  private bottomReboundEl: HTMLElement;
+  private onFocusInEvent: any;
+
+  @property({ attribute: false })
+  private hostElement = this;
+
+  connectedCallback() {
+    super.connectedCallback();
+
+    if (document.activeElement) {
+      this.previousFocus = document.activeElement;
+    }
+
+    // add attribute to prevent scroll on body
+    addAttributeValue(document.body, 'cds-layout', 'no-scrolling');
+
+    this.addReboundElements();
+
+    FocusTrapTracker.setCurrent(this.hostElement);
+    this.hostElement.setAttribute('tabindex', '0');
+    this.hostElement.focus();
+
+    this.onFocusInEvent = this.onFocusIn.bind(this);
+    document.addEventListener('focusin', this.onFocusInEvent);
+  }
+
+  disconnectedCallback() {
+    document.removeEventListener('focusin', this.onFocusInEvent);
+
+    // remove attribute that prevent scroll on body
+    removeAttributeValue(document.body, 'cds-layout', 'no-scrolling');
+
+    this.removeReboundElements();
+    FocusTrapTracker.activatePreviousCurrent();
+
+    if (isHTMLElement(this.previousFocus)) {
+      (this.previousFocus as HTMLElement).focus();
+    }
+
+    super.disconnectedCallback();
+  }
+
+  protected onFocusIn($event: any) {
+    if (FocusTrapTracker.getCurrent() !== this.hostElement) {
+      return;
+    }
+
+    if ($event.target && this.hostElement) {
+      if (
+        !this.hostElement.contains($event.target) ||
+        $event.target === this.topReboundEl ||
+        $event.target === this.bottomReboundEl
+      ) {
+        this.hostElement.focus();
+      }
+    }
+  }
+
+  protected createFocusableOffScreenEl(): any {
+    if (isBrowser()) {
+      const offScreenSpan = document.createElement('span');
+      offScreenSpan.setAttribute('tabindex', '0');
+      offScreenSpan.classList.add('offscreen-focus-rebounder');
+      return offScreenSpan;
+    }
+  }
+
+  protected addReboundElements() {
```

Comment:
With a `FocusTrapElement` interface this could be refactored out into a stand alone function 

```typescript
export function addReboundElementsToFocusTrapElement(focusTrapElement: FocusTrapElement) {
  focusTrapElement.topReboundElement = createFocusTrapReboundElement();
  focusTrapElement.bottomReboundElement = createFocusTrapReboundElement();

  if (focusTrapElement.parentElement) {
    focusTrapElement.parentElement.insertBefore(focusTrapElement.topReboundElement, focusTrapElement);
    if (focusTrapElement.nextSibling) {
      focusTrapElement.parentElement.insertBefore(focusTrapElement.bottomReboundElement, focusTrapElement.nextSibling);
    } else {
      focusTrapElement.parentElement.appendChild(focusTrapElement.bottomReboundElement);
    }
  }
}
```